src/v/cluster/rm_stm.cc

```diff
@@ -2026,9 +2027,14 @@ void rm_stm::compact_snapshot() {
 
     for (auto it = _log_state.seq_table.cbegin();
          it != _log_state.seq_table.cend();) {
-        if (
-          now.value() - it->second.entry.last_write_timestamp
-          <= expiration_ms_count) {
+        // expired producer must not have any ongoing transactions and be older
+        // than expiration threshold.
+        auto is_expired_producer
+          = now.value() - it->second.entry.last_write_timestamp
+              <= expiration_ms_count
+            && !_log_state.current_txes.contains(it->first);
```

Comment:
> !_log_state.current_txes.contains(it->first);

I think this is not a tight enough check, log_state is the state committed to the log and replayed by the stm. There might be a transaction that just began but has not committed anything to the log / there is a replication in flight.. we probably have to check something from mem_state .. @rystsov might know.

Reply:
@mmaslankaprv  check last_stable_offset()